explorer/common/trace_stream.h

```diff
@@ -23,38 +43,68 @@ namespace Carbon {
 // is fully treated as a separate file, we should be able to take a different
 // approach where the caller explicitly toggles tracing when switching file
 // contexts.
+
 class TraceStream {
  public:
   // Returns true if tracing is currently enabled.
   auto is_enabled() const -> bool {
-    return stream_.has_value() && !in_prelude_;
+    return stream_.has_value() &&
+           current_code_context_ != CodeContext::Prelude &&
+           (allowed_phases_.count(current_phase_) > 0 ||
+            allowed_phases_.count(ProgramPhase::All) > 0 ||
+            current_phase_ == ProgramPhase::Other);
   }
 
-  // Sets whether the prelude is being skipped.
-  auto set_in_prelude(bool in_prelude) -> void { in_prelude_ = in_prelude; }
-
   // Sets the trace stream. This should only be called from the main.
   auto set_stream(Nonnull<llvm::raw_ostream*> stream) -> void {
     stream_ = stream;
   }
 
+  auto set_current_phase(ProgramPhase current_phase) -> void {
+    current_phase_ = current_phase;
+  }
+
+  auto set_current_code_context(CodeContext current_code_context) -> void {
+    current_code_context_ = current_code_context;
+  }
+
+  auto set_allowed_phases(std::vector<ProgramPhase> allowed_phases) {
+    if (allowed_phases.empty()) {
+      allowed_phases_.insert(ProgramPhase::Execution);
+    } else {
+      for (auto phase : allowed_phases) {
+        allowed_phases_.insert(phase);
+      }
+    }
+  }
+
   // Returns the internal stream. Requires is_enabled.
   auto stream() const -> llvm::raw_ostream& {
     CARBON_CHECK(is_enabled());
+    CARBON_CHECK(stream_.has_value());
     return **stream_;
   }
 
+  auto current_phase() const -> ProgramPhase { return current_phase_; }
+
+  auto current_code_context() const -> CodeContext {
+    return current_code_context_;
+  }
+
   // Outputs a trace message. Requires is_enabled.
   template <typename T>
   auto operator<<(T&& message) const -> llvm::raw_ostream& {
-    CARBON_CHECK(is_enabled());
-    **stream_ << message;
+    if (is_enabled()) {
```

Comment:
Sorry, I left it there by mistake inside `operator<<`. I was using it instead of `CARBON_CHECK(is_enabled())` to debug something 
Will remove it in the next commit.

Reply:
I just found out/remembered why I did that. Do we really need to add `result` into the trace?  
https://github.com/carbon-language/carbon-lang/blob/trunk/explorer/main.cpp#L91-#95

Reply:
I don't think we need to trace the result there. If we're tracing execution, I think we should be tracing the final result of the interpreter as part of that trace instead. (And if not, we don't need to include the result value in the trace output at all.)

Reply:
Yeah, to maybe make my comment a bit clearer: I think we *should* trace the final result value if execution tracing is enabled, but that should be done by `InterpProgram`, as part of execution tracing, not here at the top level.